File: asl-app/modules/signs-hand-landmarker/android/src/main/java/expo/modules/signshandlandmarker/SignsHandLandmarkerModule.kt
package expo.modules.signshandlandmarker

import android.graphics.BitmapFactory
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.handlandmarker.HandLandmarker
import com.google.mediapipe.tasks.vision.handlandmarker.HandLandmarkerResult
import java.io.File

class SignsHandLandmarkerModule : Module() {
  private var landmarker: HandLandmarker? = null

  override fun definition() = ModuleDefinition {
    Name("SignsHandLandmarker")

    OnCreate {
      prepareLandmarker()
    }

    Function("isAvailable") {
      landmarker != null
    }

    AsyncFunction("detectFromUri") { uri: String ->
      val marker = landmarker ?: return@AsyncFunction null
      val path = if (uri.startsWith("file://")) uri.removePrefix("file://") else uri
      val file = File(path)
      if (!file.exists()) {
        return@AsyncFunction null
      }
      val bitmap = BitmapFactory.decodeFile(file.absolutePath) ?: return@AsyncFunction null
      val mpImage = BitmapImageBuilder(bitmap).build()
      val result: HandLandmarkerResult = marker.detect(mpImage)
      if (result.landmarks().isEmpty()) {
        return@AsyncFunction null
      }
      val hand = result.landmarks()[0]
      if (hand.size < 21) {
        return@AsyncFunction null
      }
      hand.take(21).map { landmark ->
        mapOf(
          "x" to landmark.x().toDouble(),
          "y" to landmark.y().toDouble(),
          "z" to landmark.z().toDouble(),
        )
      }
    }
  }

  private fun prepareLandmarker() {
    try {
      val context = appContext.reactContext ?: return
      val baseOptions = BaseOptions.builder()
        .setModelAssetPath("hand_landmarker.task")
        .build()

      val options = HandLandmarker.HandLandmarkerOptions.builder()
        .setBaseOptions(baseOptions)
        .setRunningMode(RunningMode.IMAGE)
        .setNumHands(1)
        .setMinHandDetectionConfidence(0.5f)
        .setMinHandPresenceConfidence(0.5f)
        .setMinTrackingConfidence(0.5f)
        .build()

      landmarker = HandLandmarker.createFromOptions(context, options)
    } catch (error: Exception) {
      landmarker = null
      android.util.Log.e("SignsHandLandmarker", "Init failed", error)
    }
  }
}

File: asl-app/modules/signs-hand-landmarker/ios/SignsHandLandmarkerModule.swift
import ExpoModulesCore
import MediaPipeTasksVision
import UIKit

public class SignsHandLandmarkerModule: Module {
  private var landmarker: HandLandmarker?

  public func definition() -> ModuleDefinition {
    Name("SignsHandLandmarker")

    OnCreate {
      self.prepareLandmarker()
    }

    Function("isAvailable") { () -> Bool in
      return self.landmarker != nil
    }

    AsyncFunction("detectFromUri") { (uri: String) -> [[String: Double]]? in
      guard let landmarker = self.landmarker else {
        return nil
      }

      let path = uri.hasPrefix("file://")
        ? String(uri.dropFirst("file://".count))
        : uri

      guard let image = UIImage(contentsOfFile: path) else {
        return nil
      }
      guard let mpImage = try? MPImage(uiImage: image) else {
        return nil
      }

      let result = try landmarker.detect(image: mpImage)
      guard let hand = result.landmarks.first, hand.count >= 21 else {
        return nil
      }

      return hand.prefix(21).map { landmark in
        [
          "x": Double(landmark.x),
          "y": Double(landmark.y),
          "z": Double(landmark.z),
        ]
      }
    }
  }

  private func prepareLandmarker() {
    do {
      let options = HandLandmarkerOptions()
      options.runningMode = .image
      options.numHands = 1
      options.minHandDetectionConfidence = 0.5
      options.minHandPresenceConfidence = 0.5
      options.minTrackingConfidence = 0.5

      if let modelPath = Bundle.main.path(
        forResource: "hand_landmarker",
        ofType: "task"
      ) {
        options.baseOptions.modelAssetPath = modelPath
      } else if let bundle = Bundle(identifier: "org.cocoapods.SignsHandLandmarker"),
                let modelURL = bundle.url(
                  forResource: "hand_landmarker",
                  withExtension: "task"
                ) {
        options.baseOptions.modelAssetPath = modelURL.path
      } else {
        // Fallback: search resource bundles created by CocoaPods.
        for bundle in Bundle.allBundles {
          if let modelURL = bundle.url(
            forResource: "hand_landmarker",
            withExtension: "task"
          ) {
            options.baseOptions.modelAssetPath = modelURL.path
            break
          }
        }
      }

      guard options.baseOptions.modelAssetPath != nil else {
        NSLog("[SignsHandLandmarker] Model file not found")
        return
      }

      self.landmarker = try HandLandmarker(options: options)
    } catch {
      NSLog("[SignsHandLandmarker] Init failed: \(error)")
      self.landmarker = nil
    }
  }
}
